addons/mail/models/discuss/mail_guest.py

```diff
@@ -48,9 +51,9 @@ def _compute_im_status(self):
     def _get_guest_from_context(self):
         """Returns the current guest record from the context, if applicable."""
         guest = self.env.context.get('guest')
-        if isinstance(guest, self.pool['mail.guest']):
-            return guest
-        return self.env['mail.guest']
+        if not isinstance(guest, self.pool['mail.guest']):
+            guest = self.env['mail.guest']
+        return guest.with_context(guest=guest)
```

Comment:
If guest was found from the context, why add it in the context again?

Reply:
It looks like the guest is lost on the record:

```
>>> print(self.env.context, guest.env.context)
{'lang': 'en_US', 'guest': mail.guest(16,)} {'lang': 'en_US'}
```